source/unified-test-format/unified-test-format.rst

```diff
@@ -731,6 +756,17 @@ The structure of this object is as follows:
     specification. The ``readConcern``, ``readPreference``, and ``writeConcern``
     options use the same structure as defined in `Common Options`_.
 
+.. _entity_thread_object:
+
+- ``thread``: Optional object. Defines a test runner "thread". Once the "thread"
```

Comment:
Yeah, I used "thread" in a casual way to give the general idea of what's going on here, as well as to preserve the existing wording from the old spec. In the other section, I include some other text to clarify that other concurrent execution models are just fine for implementing a thread entity:

>Test runner thread. An entity representing a "thread" that can be used to concurrently execute operations. Thread entities MUST be able to run concurrently with the main test runner thread and other thread entities, but they do not have to be implemented as actual OS threads (e.g. they can be goroutines or async tasks). See `entity_thread_object`_ for more information on how they are created.

The "threads" can kind of be thought as distinct queues of concurrent operaitions. To be "idle" essentially means there's nothing in the queue, but should an operation get added to the queue, it would begin to execute in the background / concurrently.

In Rust, the futures not making progress until they're awaited is kind of an implementation detail and doesn't really affect our model. We use async tasks to implement the "threads", which are very similar to goroutines and can run in the background without being awaited.

For Node, I could imagine a thread entity being a future, and runOnThread would basically just do something like `entityMap[threadId] = entityMap[threadId].then(() => executeOperation(o))` or something. Then `waitForThread` could be just awaiting that future. (not a node expert so sorry if this is really wrong!)

Reply:
It'd be helpful to clarify that this can be a queue of operations that get deferred until the thread is awaited. The SDAM integration tests don't apply to PHP (we rely on libmongoc for those), but that approach would be helpful if we ever do need to implement this functionality (I could see it being used for change streams perhaps).

Something else that occurred to me while discussing thread-safety: we discussed that the entity map would need to allow for concurrent access but what about the other entities within? AFAIK, some (most?) drivers don't expect threads to share MongoClient objects, and resources like cursors are certainly not thread-safe. For languages that might actually use threads or other concurrency mechanisms, we might want more guidance on what not to do in the operations sent to "thread" entities.

Reply:
> It'd be helpful to clarify that this can be a queue of operations that get deferred until the thread is awaited.

This is not the intention and implementing it this way can actually cause tests to fail. For example, imagine the ops: failCommand find, runOnThread find, waitForEvent, waitForThread. If the thread doesn't run the find() until waitForThread then the waitForEvent operation will always fail.

The "thread" is supposed to start executing the runOnThread operation immediately.

Reply:
> we might want more guidance on what not to do in the operations sent to "thread" entities.

Would a good example of this `createFindCursor` must be on the same thread as `iterateUntilDocumentOrError`? 

> This is not the intention

Good callout, so I think we should clarify that `runOnThread` should kickoff the operation not `waitOnThread`, the waiting for events is a good example of why this is important.

Reply:
> This is not the intention and implementing it this way can actually cause tests to fail. For example, imagine the ops: failCommand find, runOnThread find, waitForEvent, waitForThread. If the thread doesn't run the find() until waitForThread then the waitForEvent operation will always fail.

That is precisely why I asked if `waitForEvent` should also trigger execution of deferred operations. My line of questioning is mostly academic since PHP (with no threads) doesn't run the SDAM tests. But I'm expecting the introduction of `runOnThread` into the test runner might lead to this getting picked up by other specs, so I'll eventually need to come up with some way to handle the operation.

> I think we should clarify that `runOnThread` should kickoff the operation not `waitOnThread`, the waiting for events is a good example of why this is important.

If threads or other concurrency primitives aren't available, would executing the operations immediately as part of `runOnThread` be preferable to deferring until a `waitForX` operation?

Reply:
> I think drivers lacking concurrency primitives may just need to skip the tests

Agreed.